File: frb_example/pure_dart/lib/src/rust/api/dart_opaque_sync.dart
// ignore_for_file: invalid_use_of_internal_member, unused_import

import '../frb_generated.dart';
import 'package:flutter_rust_bridge/flutter_rust_bridge_for_generated.dart';

Object syncLoopback({required Object opaque, dynamic hint}) =>
    RustLib.instance.api.syncLoopback(opaque: opaque, hint: hint);

Object? syncOptionLoopback({Object? opaque, dynamic hint}) =>
    RustLib.instance.api.syncOptionLoopback(opaque: opaque, hint: hint);

String syncAcceptDartOpaque({required Object opaque, dynamic hint}) =>
    RustLib.instance.api.syncAcceptDartOpaque(opaque: opaque, hint: hint);

/// [DartWrapObject] can be safely retrieved on a dart thread.
String unwrapDartOpaque({required Object opaque, dynamic hint}) =>
    RustLib.instance.api.unwrapDartOpaque(opaque: opaque, hint: hint);

Object returnNonDroppableDartOpaque({required Object opaque, dynamic hint}) =>
    RustLib.instance.api
        .returnNonDroppableDartOpaque(opaque: opaque, hint: hint);

File: frb_example/pure_dart/rust/src/api/dart_opaque_sync.rs
// FRB_INTERNAL_GENERATOR: {"forbiddenDuplicatorModes": ["sync"]}

use flutter_rust_bridge::{frb, DartOpaque};

#[frb(sync)]
pub fn sync_loopback(opaque: DartOpaque) -> DartOpaque {
    opaque
}

#[frb(sync)]
pub fn sync_option_loopback(opaque: Option<DartOpaque>) -> Option<DartOpaque> {
    opaque
}

#[frb(sync)]
pub fn sync_accept_dart_opaque(opaque: DartOpaque) -> String {
    drop(opaque);
    "test".to_owned()
}

/// [DartWrapObject] can be safely retrieved on a dart thread.
#[frb(sync)]
pub fn unwrap_dart_opaque(opaque: DartOpaque) -> String {
    let handle = opaque.try_unwrap().unwrap();
    "Test".to_owned()
}

#[frb(sync)]
pub fn return_non_droppable_dart_opaque(opaque: DartOpaque) -> DartOpaque {
    let raw = opaque.try_unwrap().unwrap();
    unsafe { DartOpaque::new_non_droppable(raw.into()) }
}
